test/org/sagebionetworks/bridge/services/IntentServiceTest.java

```diff
@@ -125,40 +142,93 @@ public void submitIntentToParticipate() {
         assertEquals("Transactional", provider.getSmsType());
     }
     
-    @Test(expected = InvalidEntityException.class)
-    public void submitIntentToParticipateInvalid() {
-        IntentToParticipate intent = TestUtils.getIntentToParticipate(TIMESTAMP);
-        IntentToParticipate invalid = new IntentToParticipate.Builder().copyOf(intent).withPhone(null).build();
+    // In this case when there isn't an install link for the OS or that is marked
+    // "Universal," we take anything.
+    @Test
+    public void submitIntentToParticipateWithMismatchedInstallLinks() {
+        IntentToParticipate intent = TestUtils.getIntentToParticipate(TIMESTAMP).build();
         
-        service.submitIntentToParticipate(invalid);
+        Map<String,String> installLinks = Maps.newHashMap();
+        installLinks.put("Android", "this-is-a-link");
+        
+        when(mockStudy.getStudyIdentifier()).thenReturn(TestConstants.TEST_STUDY);
+        when(mockStudy.getInstallLinks()).thenReturn(installLinks);
+        when(mockStudy.getAppInstallLinkSmsTemplate()).thenReturn(new SmsTemplate("this-is-a-link"));
+        when(mockStudyService.getStudy(intent.getStudyId())).thenReturn(mockStudy);
+        
+        CacheKey cacheKey = CacheKey.itp(SubpopulationGuid.create("subpopGuid"), TestConstants.TEST_STUDY,
+                TestConstants.PHONE);
+        
+        service.submitIntentToParticipate(intent);
+        
+        verify(mockSubpopService).getSubpopulation(eq(mockStudy), subpopGuidCaptor.capture());
+        assertEquals(intent.getSubpopGuid(), subpopGuidCaptor.getValue().getGuid());
+        
+        verify(mockCacheProvider).setObject(keyCaptor.capture(), eq(intent), eq(4 * 60 * 60));
+        assertEquals(cacheKey, keyCaptor.getValue());
+
+        verify(mockSmsService).sendSmsMessage(isNull(), smsMessageProviderCaptor.capture());
+
+        SmsMessageProvider provider = smsMessageProviderCaptor.getValue();
+        assertEquals(mockStudy, provider.getStudy());
+        assertEquals(intent.getPhone(), provider.getPhone());
+        assertEquals("this-is-a-link", provider.getSmsRequest().getMessage());
+        assertEquals("Transactional", provider.getSmsType());        
     }
     
     @Test
-    public void submitIntentToParticipateWithoutInstallLinks() {
-        IntentToParticipate intent = TestUtils.getIntentToParticipate(TIMESTAMP);
```

Comment:
Did we accidentally delete this test?

Reply:
It seems like it. Maybe I thought installLinkCorrectlySelected() covered this, but it doesn't. Added back.